### PR TITLE
Admin Auditor

### PR DESCRIPTION
**Pull request by ec**

### CODE REVIEW

**Comment by pkucherenk0** [comment]
I have seeded next permissions to all admin roles:
```
 - { role: superadmin, verb: get, path: admin, action: audit }
 - { role: superadmin, verb: post, path: admin, action: audit }
 - { role: superadmin, verb: put, path: admin, action: audit }
 - { role: superadmin, verb: delete, path: admin, action: audit }
 - { role: superadmin, verb: get, path: api/v2, action: audit }
 - { role: superadmin, verb: post, path: api/v2, action: audit }
 - { role: superadmin, verb: put, path: api/v2, action: audit }
 - { role: superadmin, verb: delete, path: api/v2, action: audit }
```

By calling to  `{{barong}}admin/activities/admin`  was able to get **only** activity related to `"action": "list"` (that`s requesting metrics, list of users ...) and it was not consistent. 

Creating role with api  `{{barong}}admin/permissions` had the same zero effect on admin activity logs.

**Comment by ec** [comment]
@pkucherenk0 let's pair on it at Monday and check together.
Probably, I will need to add and will add a MD file explaining the whole process

**Comment by pkucherenk0** [comment]
1. Fix the seed.rb so we permissions with same role, verb and path but with different action and topic was seeded.
2. Truck uuid_target with post request

**Review by calj — APPROVED** [review]

File: db/schema.rb [hunk 1]
```diff
@@ -10,17 +10,20 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2019_03_18_133453) do
+ActiveRecord::Schema.define(version: 2019_05_29_114214) do
 
   create_table "activities", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8", force: :cascade do |t|
     t.bigint "user_id", null: false
+    t.string "target_uuid"
```
Comment (mod) [inline db/schema.rb]: t.string "source_uid"
t.string "target_uid"

File: lib/barong/seed.rb [hunk 2]
```diff
@@ -45,7 +45,7 @@ def seed_permissions
       logger.info "Seeding permissions"
       seeds["permissions"].each do |perm|
         logger.info "---"
-        if Permission.find_by(role: perm["role"], verb: perm["verb"], path: perm["path"]).present?
+        if Permission.find_by(role: perm["role"], verb: perm["verb"], path: perm["path"], action: perm["action"]).present?
```
Comment (mod) [inline lib/barong/seed.rb:48]: convert all into 5 inserts

File: lib/barong/authorize.rb [hunk 3]
```diff
@@ -69,8 +71,33 @@ def validate_permissions!(user)
...
+      end
+    end
 
-      error!({ errors: ['authz.invalid_permission'] }, 401) if permissions.blank? || permissions.pluck(:action).include?('DROP')
+    def log_activity(user_id, result, topic)
+      ActivityLogger.write(
+        user_id: user_id,
+        result: result,
+        user_agent: @request.env['HTTP_USER_AGENT'],
+        user_ip: @request.ip,
```
Comment (mod) [inline lib/barong/authorize.rb]: this ip is the ip from proxy
XFORWARDED_FOR

File: lib/barong/authorize.rb [hunk 4]
```diff
@@ -69,8 +71,33 @@ def validate_permissions!(user)
         Rails.cache.write('permissions', permissions, expires_in: 5.minutes)
       end
       permissions.select! { |a| a.role == user.role && a.verb == @request.env['REQUEST_METHOD'] && @path.starts_with?(a.path) }
+      actions = permissions.blank? ? [] : permissions.pluck(:action).uniq
+
+      if permissions.blank? || actions.include?('DROP') || !actions.include?('ACCEPT')
+        if actions.include?('AUDIT')
```
Comment (mod) [inline lib/barong/authorize.rb]: ALWAYS LOG DROP and permissions denied with result "denied"

**Review by mod — CHANGES_REQUESTED** [review]

**Review by mod — APPROVED** [review]

**Pull request merged**